src/main/rascal/lang/jimple/analysis/saa/DominanceTree.rsc

```diff
@@ -1,15 +1,45 @@
 module lang::jimple::analysis::saa::DominanceTree
 
-import List;
+import demo::Dominators;
+import Set;
+import Relation;
 import lang::jimple::analysis::FlowGraph;
+import analysis::graphs::Graph;
 
-public map[Node, Node] createDominanceTree(FlowGraph flowGraph) {
-	result = (computeDominator(flowGraph, a): a | <a, _> <- flowGraph);
-	return result;
+/*
+	This tree returned is a set of dominance subtrees for each node, the ideal tree would be a one
+	that contains the the immediate childs for each leaf instead all the dominated ones.
+	
+	Once done, the findIdom function will be far more simple, because we would only need
+	to find for leaf that contains a given child. So that leaf would be the immediate domminator.
+*/
+
+public map[&T, set[&T]] createDominanceTree(Graph[&T] graph) {
+	PRED = graph;
+	ROOT = entryNode();
+	
+	set[&T] rootDominators = reachX(PRED, {ROOT}, {});
+	set[&T] VERTICES = carrier(PRED);
+
+	return (V: (rootDominators - reachX(removeNodeFromGraph(graph, V), {ROOT}, {V}) - {V}) | &T V <- VERTICES );
+}
+
+public Graph[&T] removeNodeFromGraph(Graph[&T] graph, &T nodeToRemove) {
+ return { <father, graphChild> | <father, graphChild> <- graph, nodeToRemove != father || nodeToRemove == entryNode() };
 }
 
-public Node computeDominator(FlowGraph flowGraph, Node currentNode) {
-	predecessorNodeList = [a | <a, b> <- flowGraph, b == currentNode];
-	result = if(isEmpty(predecessorNodeList) == true) currentNode; else head(predecessorNodeList);
-	return result;
+
+public Node findIdom(map[&T, set[&T]] dominanceTree, Node child) {
+	ROOT = entryNode();
+	idom = entryNode();
+	
+	possibleIdoms = [ father | father <- dominanceTree, child in dominanceTree[father] ];
+
+	for(possibleIdom <- possibleIdoms) {
+		if(size(dominanceTree[possibleIdom]) < size(dominanceTree[idom])) {
+			idom = possibleIdom;
+		};
+	};
+	
+	return idom;
```

Comment:
Use the size to determinate them nearest idom

Reply:
Also based on https://homepages.cwi.nl/~jurgenv/papers/GTTSE-2010.pdf and https://stackoverflow.com/questions/64380865/dominator-tree-of-a-rascal-graph/64385252